rust/datafusion/src/physical_plan/unicode_expressions.rs

```diff
@@ -0,0 +1,537 @@
+
+                            if length < graphemes.len() {
+                                Some(graphemes[..length].concat())
+                            } else if fill_chars.is_empty() {
+                                Some(string.to_string())
+                            } else {
+                                let mut s = string.to_string();
+                                let mut char_vector =
+                                    Vec::<char>::with_capacity(length - graphemes.len());
+                                for l in 0..length - graphemes.len() {
+                                    char_vector.push(
+                                        *fill_chars.get(l % fill_chars.len()).unwrap(),
+                                    );
+                                }
+                                s.insert_str(
+                                    0,
+                                    char_vector.iter().collect::<String>().as_str(),
+                                );
+                                Some(s)
+                            }
+                        }
+                    }
+                })
+                .collect::<GenericStringArray<T>>();
+
+            Ok(Arc::new(result) as ArrayRef)
+        }
+        other => Err(DataFusionError::Internal(format!(
+            "lpad was called with {} arguments. It requires at least 2 and at most 3.",
+            other
+        ))),
+    }
+}
+
+/// Reverses the order of the characters in the string.
+/// reverse('abcde') = 'edcba'
+pub fn reverse<T: StringOffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = downcast_string_arg!(args[0], "string", T);
+
+    let result = string_array
+        .iter()
+        .map(|string| {
+            string.map(|string: &str| string.graphemes(true).rev().collect::<String>())
+        })
+        .collect::<GenericStringArray<T>>();
+
+    Ok(Arc::new(result) as ArrayRef)
+}
+
+/// Returns last n characters in the string, or when n is negative, returns all but first |n| characters.
+/// right('abcde', 2) = 'de'
+pub fn right<T: StringOffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = downcast_string_arg!(args[0], "string", T);
+    let n_array = downcast_arg!(args[1], "n", Int64Array);
+
+    let result = string_array
+        .iter()
+        .zip(n_array.iter())
+        .map(|(string, n)| match (string, n) {
+            (None, _) => None,
+            (_, None) => None,
+            (Some(string), Some(n)) => match n.cmp(&0) {
+                Ordering::Equal => Some(""),
+                Ordering::Greater => Some(
+                    string
+                        .grapheme_indices(true)
+                        .rev()
+                        .nth(n as usize - 1)
+                        .map_or(string, |(i, _)| {
+                            &from_utf8(&string.as_bytes()[i..]).unwrap()
```

Comment:
here is another use of unwrap because of trying to find grapheme boundaries based on `grapheme_indicies` -- if we could use character length math rather than byte boundaries we might be able to avoid the unwraps